### PR TITLE
🐳 feat(protocol): alpha-4 with staking-based tokenomics

### PR DESCRIPTION
**Pull request by adaki2004**
// New concept TLDR:
    // 2 main data 'registry':
    // - TopProver array: we keep track of the top32. We need an array because we need to iterate over
    // - ExitingProver mapping: each and every prover who is either exiting or kicked out (by someone staking more) is
    // here until they are not withdraw the funds (with exit())

    // NOTE:
    // - signalling an exit if you are a prover can be done via stake(totalAMount = 0)
    // - modifying existing parameters also done via stake (see comments there)
    // - there are mechanisms in the stake() code to handle scenarios like:
    //     A: Bob is the least staker. Alice comes in and stakes more than Bob and Carol.
    //     Now Bob forced to leave and not within the ExitingProvers. He realizes it and
    //     tries to re-stake. Now he stakes more than Carol, so Carol is gone, but Bob can
    //     re-use his funds currently in the 'staking queue'. So basically he is out of the
    //     exiting queue.
    //
    //     B: Almost same situation as above, except that when Bob is leaving (force leaving)
    //     Carol decided to leave as well, so she is not in the topProver array anymore. It means
    //     Bob could come back with his stake (currenlty in the exit 'queue') or even less than that
    //     amount since Carol left, there is 1 "free space". So he could decide to re-stake less than
    //     his amount in the current exit 'queue', so technically he still have some left in the exit queue
    //     while also being a topProver in the same time!

### CODE REVIEW

**Comment by cyberhorsey** [comment]
i think we can remove `systemProofCooldownPeriod` from config now, right?

File: packages/protocol/contracts/L1/TaikoConfig.sol [hunk 1]
```diff
@@ -11,47 +11,45 @@ import { TaikoData } from "../L1/TaikoData.sol";
 library TaikoConfig {
     function getConfig() internal pure returns (TaikoData.Config memory) {
         return TaikoData.Config({
+            // Group 1: general configs
```
Comment (dantaik) [inline packages/protocol/contracts/L1/TaikoConfig.sol:14]: please run `pnpm fmt:sol`

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+        uint32 feePerGas
+    )
+        external
+        returns (address prover, uint32 rewardPerGas);
+
+    function releaseProver(address prover) external;
+    function slashProver(address prover) external;
+
+    // TODO: we need to add this as well.
+    function getAvailableCapacity() external returns (uint256);
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:25]:
```suggestion
    function getAvailableCapacity() external view returns (uint256);
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+    // 2 main data 'registry':
+    // - TopProver array: we keep track of the top32. We need an array because we need to iterate over
+    // - ExitingProver mapping: each and every prover who is either exiting or kicked out (by someone staking more) is
+    // here until they are not withdraw the funds (with exit())
+
+    // NOTE:
+    // - signalling an exit if you are a prover can be done via stake(totalAMount = 0)
+    // - modifying existing parameters also done via stake (see comments there)
+    // - there are mechanisms in the stake() code to handle scenarios like:
+    //     A: Bob is the least staker. Alice comes in and stakes more than Bob and Carol.
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:39]: please run `pnpm fmt:sol` then fix the layout of these comments

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+    //     his amount in the current exit 'queue', so technically he still have some left in the exit queue
+    //     while also being a topProver in the same time!
+
+
+    // This is 1/4 slot
+    // TODO: Dani, this structure should hold all information required to
+    // dynamically calculate the weight of the top 32 provers and select one
+    // based on a random number.
+    struct TopProver {
+        uint32 amount; // this value will change when we slash the prover
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:58]:
```suggestion
        uint32 amount; // will slash from ExistingProver's exitAmount then from this value.
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        uint16 rewardPerGas;
+        uint16 currentCapacity;
+    }
+
+    // prover always have one Prover object,regardless if it is active or exiting.
+    // @TODO: We do not need to use both. We shall have 1 for TopProvers, and 1 for ExitingProvers
+    struct ExitingProver {
+        uint64 exitRequestedAt;
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:66]: If we name this structure as `ExitingProver`, we can remove the 'exit" prefix from its variables.

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+        uint32 exitAmount;
+        // @QUESTION: I think 'usedCapacity' is not needed here or at least cannot be calculated only from the
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:68]: If Alice staked with capacity=200, now 150 of them are used, Alice requested to exit, (now we lost all capacity information), then Alice staked again with capacity=500, now the 150 blocks are slowly returned, then the capacity will Alice will likely increase to more than 500 capacity, how to handle it?

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+
+    // Id mappings
+    mapping(uint8 id => address) idToProver;
+    mapping(address prover => uint8 id) proverToId;
+
+    // Exit period. Might fine tune later on !
+    uint256 public constant EXIT_PERIOD = 1 weeks;
+
+    // 500 means 5% if 10_000 is 100 %
+    uint256 public constant SLASH_AMOUNT_IN_BP = 500;
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:90]:
```suggestion
    uint256 public constant SLASH_AMOUNT_IN_BP = 500;  // basis points
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+    mapping(uint8 id => address) idToProver;
+    mapping(address prover => uint8 id) proverToId;
+
+    // Exit period. Might fine tune later on !
+    uint256 public constant EXIT_PERIOD = 1 weeks;
+
+    // 500 means 5% if 10_000 is 100 %
+    uint256 public constant SLASH_AMOUNT_IN_BP = 500; 
+
+    uint256 public WEI_TO_1_TKO = 10e8;
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:92]:
```suggestion
    uint256 public ONE_TKO = 10e8;
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+
+    uint256[100] private __gap;
+
+    event ProverEntered(
+        address prover, uint32 amount, uint16 rewardPerGas, uint16 capacity
+    );
+
+    event ProverChangedParameters(address prover, uint32 newBalance, uint16 newReward, uint16 newCapacity);
+
+    event ProverKickeOutByWithAmount(address kickedOut, address newProver, uint32 totalAmount);
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:102]: We do not need ProverKickeOutByWithAmount, 
We can use two events, ProverExited + ProverStaked.

I suggest to remove "Prover" prefix, simply use the following events:

Staked/Exiting/Exited/Slashed

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+    event ProverExitRequested(address prover, uint64 timestamp);
+
+    event ProverExited(address prover, uint64 timestamp);
+
+    event ProverSlashed(address prover, uint32 newBalance);
+
+    modifier onlyProtocol() {
+        require(
+            AddressResolver(this).resolve("taiko", false) != msg.sender,
+            "Only Taiko L1 can call this function"
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:113]: custom error

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+
+    event ProverKickeOutByWithAmount(address kickedOut, address newProver, uint32 totalAmount);
+
+    event ProverExitRequested(address prover, uint64 timestamp);
+
+    event ProverExited(address prover, uint64 timestamp);
+
+    event ProverSlashed(address prover, uint32 newBalance);
+
+    modifier onlyProtocol() {
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:110]:
```suggestion
    modifier onlyFromProtocol() {
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+            AddressResolver(this).resolve("taiko", false) != msg.sender,
+            "Only Taiko L1 can call this function"
+        );
+        _;
+    }
+
+    modifier onlyActiveProver(address prover) {
+        // Either within top provers (rewardPerGas != 0 serves as a check), or still exiting
+        require(
+            topProvers[proverToId[prover]].rewardPerGas != 0
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:121]: I suggest for the first prover, the id is not `0`, it should be `1`, then here we can simply check

```solidity
proverToId[prover] != 0
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 13]
```diff
@@ -0,0 +1,479 @@
...
+            || 
+            ( exitingProvers[prover].exitRequestedAt != 0
+            && exitingProvers[prover].exitRequestedAt + EXIT_PERIOD < block.timestamp),
+            "Only active provers can call this function"
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:125]: Again. this comment is not right, we are not checking msg.sender

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 14]
```diff
@@ -0,0 +1,479 @@
...
+        );
+        _;
+    }
+
+    modifier onlyActiveProver(address prover) {
+        // Either within top provers (rewardPerGas != 0 serves as a check), or still exiting
+        require(
+            topProvers[proverToId[prover]].rewardPerGas != 0
+            || 
+            ( exitingProvers[prover].exitRequestedAt != 0
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:123]: why we check existing prover here if the modifier name. is `onlhyActiveProver`?

I strongly suggest you remove this modifier and bring the logics into the function, this modifier is used only once after all.

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 15]
```diff
@@ -0,0 +1,479 @@
...
+     * Initialize the rollup.
+     *
+     * @param _addressManager The AddressManager address.
+     */
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    // A demo how to optimize the assignProver by using only 8 slots. It's still
+    // a lot of slots tough.
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:140]:
```suggestion
    // a lot of slots though.
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 16]
```diff
@@ -0,0 +1,479 @@
...
+    function assignProver(
+        uint64 blockId,
+        uint32 feePerGas
+    )
+        external
+        returns (address prover, uint32 rewardPerGas)
+    {
+        // readjust each prover's rate
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:148]:
```suggestion
        // calculate each prover's dynamic weight
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 17]
```diff
@@ -0,0 +1,479 @@
...
+        uint256[32] memory weights;
+        uint256 totalWeight;
+
+        // Iterate over in-memory array
+        TopProver[32] memory mTopProvers = topProvers;
+
+        for (uint8 i; i < mTopProvers.length; ++i) {
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:155]: break the loop once an prover's rewardPerGas is 0

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 18]
```diff
@@ -0,0 +1,479 @@
...
+    )
+        external
+        returns (address prover, uint32 rewardPerGas)
+    {
+        // readjust each prover's rate
+        uint256[32] memory weights;
+        uint256 totalWeight;
+
+        // Iterate over in-memory array
+        TopProver[32] memory mTopProvers = topProvers;
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:153]: We should load topProvers one by one, not all at once, assuming there is only one prover in the list...

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 19]
```diff
@@ -0,0 +1,479 @@
...
+
+        if (totalWeight == 0) {
+            return (address(0), 0);
+        }
+
+        // Determine prover idx
+        bytes32 rand =
+            keccak256(abi.encode(blockhash(block.number - 1), blockId));
+
+        uint8 id = _pickProver({
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:168]: I also suggest to inline the _pickProver as it is small and used only here

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+        onlyProtocol
+        onlyActiveProver(prover)
+    {
+        // We need to determine if the prover is an exiting prover or normal (top)
+        // It might happen that prover is in the exit, while also in the top, in such case
+        // try to punish the exiting amount
+        uint32 afterSlash;
+        if (
+            exitingProvers[prover].exitRequestedAt != 0
+                && exitingProvers[prover].exitRequestedAt + EXIT_PERIOD < block.timestamp
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:202]:
```suggestion
         
```
// no need to check this!

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 21]
```diff
@@ -0,0 +1,479 @@
...
+        external
+        onlyProtocol
+        onlyActiveProver(prover)
+    {
+        // We need to determine if the prover is an exiting prover or normal (top)
+        // It might happen that prover is in the exit, while also in the top, in such case
+        // try to punish the exiting amount
+        uint32 afterSlash;
+        if (
+            exitingProvers[prover].exitRequestedAt != 0
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:201]:
```suggestion
            exitingProvers[prover].exitAmount != 0
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 22]
```diff
@@ -0,0 +1,479 @@
...
+                && exitingProvers[prover].exitRequestedAt + EXIT_PERIOD < block.timestamp
+        ) {
+            // Exiting prover get's slashed
+            uint32 currentStaked = exitingProvers[prover].exitAmount;
+
+            afterSlash = uint32(
+                currentStaked * uint256(10_000 - SLASH_AMOUNT_IN_BP) / 10_000
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:208]: I would do this:

```solidity
uint32 amountToExit = existingProvers[prover].amount
uint32 amountToSlash = (topProver[id].amount + amountToExit) * percentage.

if (amountToExit >=  amountToSlash) {
    existingProvers[prover].amount  = amountToExit - amountToSlash;
} else {
    if (amountToExit > 0) {
      delete  existingProvers[prover];
      amountToSlash -= amountToExit;
    } 
    topProver[id].amount -= amountToSlash;
}
```

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 23]
```diff
@@ -0,0 +1,479 @@
...
+    function stake(
+        uint32 totalAmount,
+        uint16 rewardPerGas,
+        uint16 capacity
+    )
+        external
+        nonReentrant
+    {
+
+        address newProver = _msgSender();
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:242]: Do not use `_msgSender`

File: packages/protocol/contracts/L1/ProverPool.sol [hunk 24]
```diff
@@ -0,0 +1,479 @@
...
+        if(totalAmount == 0) {
```
Comment (dantaik) [inline packages/protocol/contracts/L1/ProverPool.sol:243]: totalAmount is 0 is an full exit, but if totalAmount is 1. then it's still an exit. The code should handle them almost the same...

File: packages/protocol/contracts/L1/libs/LibProposing.sol [hunk 25]
```diff
@@ -73,37 +97,55 @@ library LibProposing {
...
         }
 
-        emit BlockProposed(state.numBlocks, meta, blockFee);
+        IMintableERC20(resolver.resolve("taiko_token", false)).burn({
+            from: msg.sender,
+            amount: getBlockFee(state, config, meta.gasLimit)
+        });
+
+        emit BlockProposed(state.numBlocks, meta);
+
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/libs/LibProposing.sol:147]: Lets keep it as-is if it works, then try to optimize the overall gas later.

**Pull request merged**